### PR TITLE
apiextensions-apiserver: TestFinaliazationAndDeletion integration test

### PR DESCRIPTION
**Pull request by xmudrii**
**What this PR does / why we need it**: This PR adds an integration test for [behavior described in the issue #60538 (comment)](https://github.com/kubernetes/kubernetes/issues/60538#issuecomment-369182483).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Partially addresses #60538 -- tests

**Special notes for your reviewer**: This is based of the PR #60542, so this shouldn't be merged before that one. Tests are failing on the line 155, due to the timeout/non-nil error. I've tried to debug it furtherly, and found out that the CRD is still terminating. 

Actually, the CRD gets deleted, but it takes some time. For example, if you add `time.Sleep(x * time.Second)` on the line 152, the test would pass. The problem is this isn't predictable. With `x` of 6-10 seconds, I manage to get this test passing, but there's zero guarantees the CRD will get deleted for that 10 seconds.

I have found the `wait.Poll` function in some tests, but I'm not sure can it fix the problem I have.

This is my first PR to the Kubernetes project, so I could be missing something. Any idea is appreciated. :)

**Release note**:

```release-note
NONE
```

/sig api-machinery
/area custom-resources
/cc nikhita sttts liggitt

### CODE REVIEW

**Comment by xmudrii** [comment]
@k8s-ci-robot CLA signed

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go [hunk 1]
```diff
@@ -89,3 +89,68 @@ func TestFinalization(t *testing.T) {
...
+		if !errors.IsConflict(err) {
+			require.NoError(t, err) // Fail on unexpected error
+		}
+		gottenNoxuInstance, err = noxuResourceClient.Get(name, metav1.GetOptions{})
+		require.NoError(t, err)
+	}
+
+	// Verify the CR is gone.
+	// It should return the NonFound error.
+	gottenNoxuInstance, err = noxuResourceClient.Get(name, metav1.GetOptions{})
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: this needs a wait.Until loop. The deletion might need a second.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go [hunk 2]
```diff
@@ -89,3 +89,68 @@ func TestFinalization(t *testing.T) {
...
+	require.Error(t, err)
+
+	// Verify is the CRD gone.
+	_, err = testserver.GetCustomResourceDefinition(noxuDefinition, apiExtensionClient)
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: also this needs a wait loop.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go [hunk 3]
```diff
@@ -89,3 +89,68 @@ func TestFinalization(t *testing.T) {
...
+
+	// Delete the CRD.
+	testserver.DeleteCustomResourceDefinition(noxuDefinition, apiExtensionClient)
+
+	// Check is CR still there after the CRD deletion.
+	gottenNoxuInstance, err = noxuResourceClient.Get(name, metav1.GetOptions{})
+	require.NoError(t, err)
+
+	// Update the CR to remove the finalizer.
+	for {
```
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: why is this needed?

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go [hunk 4]
```diff
@@ -89,3 +89,68 @@ func TestFinalization(t *testing.T) {
...
+		gottenNoxuInstance.SetFinalizers(nil)
+		_, err = noxuResourceClient.Update(gottenNoxuInstance)
```
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: The update should happen instantly.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go (hunk 3 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: the might be conflict in general (maybe not in this example). So this loop make sense IMO.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go [hunk 5]
```diff
@@ -89,3 +89,68 @@ func TestFinalization(t *testing.T) {
...
+			require.NoError(t, err) // Fail on unexpected error
+		}
+		gottenNoxuInstance, err = noxuResourceClient.Get(name, metav1.GetOptions{})
+		require.NoError(t, err)
+	}
+
+	// Verify the CR is gone.
+	// It should return the NonFound error.
+	gottenNoxuInstance, err = noxuResourceClient.Get(name, metav1.GetOptions{})
+	require.Error(t, err)
```
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: For checking if the error is `NotFound`, you can use `apierrors.IsNotFound(err)` where `apierrors` is `"k8s.io/apimachinery/pkg/api/errors"`

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go (hunk 2 above)
Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: You can use `apierrors.IsNotFound` in the wait loop as well. :)

**Comment by xmudrii** [comment]
@sttts @nikhita I'm not sure am I doing this right, but tests are passing now.
Does it need any timeout function or tests will timeout automatically?

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go [hunk 6]
```diff
@@ -89,3 +92,78 @@ func TestFinalization(t *testing.T) {
...
+			require.NoError(t, err) // Fail on unexpected error
+		}
+		gottenNoxuInstance, err = noxuResourceClient.Get(name, metav1.GetOptions{})
+		require.NoError(t, err)
+	}
+
+	waitCh := make(chan struct{})
+	// Verify the CR is gone.
+	// It should return the NonFound error.
+	err = wait.PollUntil(500*time.Millisecond, func() (bool, error) {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: Something like `wait.Poll(50*time.Millisecond, wait.ForeverTestTimeout, ...` is used elsewhere in tests.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go [hunk 7]
```diff
@@ -89,3 +92,78 @@ func TestFinalization(t *testing.T) {
...
+		_, err = noxuResourceClient.Get(name, metav1.GetOptions{})
+		return apierrors.IsNotFound(err), err
+	}, waitCh)
+	if !apierrors.IsNotFound(err) {
+		t.Fatal(err)
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: add some context what failed.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go (hunk 6 above)
Comment (xmudrii) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: Changed :)

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go (hunk 7 above)
Comment (xmudrii) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: Added

**Comment by xmudrii** [comment]
@sttts Addressed comments from the latest reviews. Tests are passing for me :green_apple:

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go [hunk 8]
```diff
@@ -89,3 +92,77 @@ func TestFinalization(t *testing.T) {
...
+	}
+
+	// Verify the CR is gone.
+	// It should return the NonFound error.
+	err = wait.Poll(500*time.Millisecond, wait.ForeverTestTimeout, func() (bool, error) {
+		_, err = noxuResourceClient.Get(name, metav1.GetOptions{})
+		return apierrors.IsNotFound(err), err
+	})
+	if !apierrors.IsNotFound(err) {
+		t.Fatalf("unable to delete noxuResourceClient: %v", err)
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: a bit misleading, the `noxuResourceClient` is not deleted.

Comment (xmudrii) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: is something like this better?

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: yes, perfect

**Comment by sttts** [comment]
Lgtm.

Waiting for #60542 to merge.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go [hunk 9]
```diff
@@ -89,3 +92,77 @@ func TestFinalization(t *testing.T) {
...
+		if !errors.IsConflict(err) {
+			require.NoError(t, err) // Fail on unexpected error
+		}
+		gottenNoxuInstance, err = noxuResourceClient.Get(name, metav1.GetOptions{})
+		require.NoError(t, err)
+	}
+
+	// Verify the CR is gone.
+	// It should return the NonFound error.
+	err = wait.Poll(500*time.Millisecond, wait.ForeverTestTimeout, func() (bool, error) {
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: Why does it need a Poll here? If the finalizer is removed, the CR should be deleted immediately.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go [hunk 10]
```diff
@@ -89,3 +92,77 @@ func TestFinalization(t *testing.T) {
...
+		_, err = noxuResourceClient.Get(name, metav1.GetOptions{})
+		return apierrors.IsNotFound(err), err
+	})
+	if !apierrors.IsNotFound(err) {
+		t.Fatalf("unable to delete cr: %v", err)
+	}
+
+	err = wait.Poll(500*time.Millisecond, wait.ForeverTestTimeout, func() (bool, error) {
+		_, err = testserver.GetCustomResourceDefinition(noxuDefinition, apiExtensionClient)
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: Just curious, is the deletion of CRD blocked by its CRs?

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go (hunk 9 above)
Comment (xmudrii) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: As far as I understood, it can take few seconds sometimes to get delete. [More details](https://github.com/kubernetes/kubernetes/pull/60592#discussion_r171468946).

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go (hunk 10 above)
Comment (xmudrii) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: I'm not sure as I'm a first time contributor, but I think it is.

**Comment by xmudrii** [comment]
I've removed the WIP label and rebased against the master branch, as #60538 is merged.

/assign sttts

**Comment by xmudrii** [comment]
I'm not sure why this happens -- that's not really related to my tests. Is there something I should fix?

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go (hunk 10 above)
Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: > Just curious, is the deletion of CRD blocked by its CRs?

Yes.

**Comment by nikhita** [comment]
> I'm not sure why this happens -- that's not really related to my tests. Is there something I should fix?

No, it's not your tests. :)

Sometimes tests might [flake](https://github.com/kubernetes/community/blob/master/contributors/devel/flaky-tests.md) and the best way to fix them is to do a

/retest

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go (hunk 9 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: @caesarxuchao do we guarantee that? From the Finalizer field godoc I only ready that it might be deleted. I don't see that it guaranteed to be deleted right away synchronously in the UPDATE request. Moreover, HA servers might still see it from their caches.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go (hunk 10 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: There is a "terminating" condition on the CRD. New CRs are blocked from being created. And there is a controller deleting the existing ones.

**Comment by sttts** [comment]
/approve
Lgtm. Waiting for @caesarxuchao on https://github.com/kubernetes/kubernetes/pull/60592#discussion_r171697341.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go (hunk 9 above)
Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: @caesarxuchao Does this lgty?

Comment (caesarxuchao) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: If the Update removes the last finalizer, the generic registry deletes the object synchronously (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L631). The CR registry inherits the generic registry, so that's the expected behavior of CR as well. Also, this particular test isn't using the HA setup, so I think we don't need the poll.

If you agree, we can send a followup PR to update field godoc.

**Comment by xmudrii** [comment]
‎Should I remove the poll now, or we can get this in as is, and then later submit PR to address needed issues?

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go (hunk 9 above)
Comment (xmudrii) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: @caesarxuchao @nikhita It seems like the test is passing even without that poll. Should I remove it from the test?

Comment (caesarxuchao) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: Yes, please remove. Then lgtm.

Comment (xmudrii) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go]: @caesarxuchao @nikhita Done :)

**Comment by nikhita** [comment]
@xmudrii Congratulations on getting your first PR merged! :tada:
